File: backend/worker/frame.test.ts
import { test, expect } from "bun:test";
import {
  encodeFrame,
  FrameReader,
  encodeResponse,
  decodeRequest,
  KIND_RENDER_REQUEST,
  KIND_RENDER_RESPONSE,
  RAW_FRAME_SIZE,
  MAX_FRAME,
  MAX_REQUEST_FRAME,
  MAX_REQUEST_IMAGE_BYTES,
  META_MAX,
  type ResponseMeta,
} from "./frame";

test("constants match the faasproto spec (proto v2 direction-split ceilings)", () => {
  expect(RAW_FRAME_SIZE).toBe(360000);
  expect(META_MAX).toBe(16384);
  expect(MAX_FRAME).toBe(1 + 4 + 16384 + 360000);
  // K7: the request ceiling is larger (multi-MB base64 source image) and must exceed the response one.
  expect(MAX_REQUEST_FRAME).toBe(1 + 16384 + (Math.floor(MAX_REQUEST_IMAGE_BYTES / 3) + 1) * 4);
  expect(MAX_REQUEST_FRAME).toBeGreaterThan(MAX_FRAME);
});

test("encodeFrame produces the spec wire format (u32be total | kind | payload)", () => {
  const f = encodeFrame(KIND_RENDER_REQUEST, new Uint8Array([0x68, 0x69])); // "hi"
  expect(Array.from(f)).toEqual([0x00, 0x00, 0x00, 0x03, 0x01, 0x68, 0x69]);
});

test("FrameReader round-trips and reassembles across chunk boundaries", () => {
  const payload = new TextEncoder().encode("a".repeat(1000));
  const frame = encodeFrame(KIND_RENDER_RESPONSE, payload);
  const r = new FrameReader();
  // feed in three arbitrary splits
  expect(r.feed(frame.slice(0, 2))).toHaveLength(0);
  expect(r.feed(frame.slice(2, 500))).toHaveLength(0);
  const out = r.feed(frame.slice(500));
  expect(out).toHaveLength(1);
  expect(out[0].kind).toBe(KIND_RENDER_RESPONSE);
  expect(out[0].payload).toEqual(payload);
});

test("FrameReader yields multiple frames from one chunk", () => {
  const a = encodeFrame(0x01, new Uint8Array([1]));
  const b = encodeFrame(0x02, new Uint8Array([2, 3]));
  const merged = new Uint8Array(a.length + b.length);
  merged.set(a, 0);
  merged.set(b, a.length);
  const out = new FrameReader().feed(merged);
  expect(out.map((f) => f.kind)).toEqual([0x01, 0x02]);
});

// K7: FrameReader reads the sup→worker REQUEST direction, so its ceiling is MAX_REQUEST_FRAME. A
// length that fit the old single MAX_FRAME ceiling is now accepted (the request may be that large);
// only a length over MAX_REQUEST_FRAME is rejected before a body is read. (The old MAX_FRAME+1 assert
// broke on purpose — that break is the TS half of the K7 red proof.)
test("FrameReader admits a request up to MAX_REQUEST_FRAME and rejects beyond it", () => {
  // A former-oversize length (MAX_FRAME+1) is now within the request ceiling: NOT rejected on the prefix.
  const under = new Uint8Array(4);
  new DataView(under.buffer).setUint32(0, MAX_FRAME + 1, false);
  expect(new FrameReader().feed(under)).toHaveLength(0); // waits for the body, no throw
  // Over the request ceiling → rejected before reading a body.
  const over = new Uint8Array(4);
  new DataView(over.buffer).setUint32(0, MAX_REQUEST_FRAME + 1, false);
  expect(() => new FrameReader().feed(over)).toThrow(/> max/);
});

// K7 (response wall preserved): encodeFrame writes the worker→sup RESPONSE direction and stays tight
// at MAX_FRAME — the larger request ceiling must not leak into the direction the untrusted worker emits.
test("encodeFrame rejects a response payload over the tight MAX_FRAME ceiling", () => {
  expect(() => encodeFrame(KIND_RENDER_RESPONSE, new Uint8Array(MAX_FRAME))).toThrow(/> max/);
});

test("encodeResponse enforces the fixed 360000-byte image invariant", () => {
  const raw = new Uint8Array(RAW_FRAME_SIZE);
  const ok: ResponseMeta = { v: 1, id: "x", ok: true, w: 400, h: 300, err: null };
  expect(encodeResponse(ok, raw).length).toBe(4 + JSON.stringify(ok).length + RAW_FRAME_SIZE);
  // ok response with the wrong image size is rejected
  expect(() => encodeResponse(ok, raw.subarray(0, RAW_FRAME_SIZE - 1))).toThrow(/!=/);
  // error response must carry no image
  const err: ResponseMeta = { v: 1, id: "x", ok: false, w: 400, h: 300, err: { kind: "throw", msg: "boom" } };
  expect(() => encodeResponse(err, raw)).toThrow(/no image/);
  expect(encodeResponse(err, null).length).toBeGreaterThan(4);
});

test("decodeRequest parses a request payload", () => {
  const req = {
    v: 1,
    id: "abc",
    fn: { id: 1, ver: 1, source: "x" },
    ctx: { serial: "S1", channel: "c", trigger: { type: "render" }, now: "2026-07-01T00:00:00Z" },
    force: false,
    dither_default: "none",
    secrets: { tok: "V" },
    egress_allow: ["ha.local:8123"],
    egress_cred: "cred",
    limits: { timeout_ms: 8000, mem_mb: 256 },
  };
  const got = decodeRequest(new TextEncoder().encode(JSON.stringify(req)));
  expect(got.ctx.serial).toBe("S1");
  expect(got.secrets.tok).toBe("V");
});

File: backend/internal/faasproto/frame_test.go
package faasproto

import (
	"bytes"
	"encoding/binary"
	"encoding/json"
	"strings"
	"testing"
)

func TestFrameRoundTrip(t *testing.T) {
	for _, kind := range []byte{KindRenderRequest, KindRenderResponse, KindCancel, KindReady} {
		payload := []byte("payload-for-" + string([]byte{kind}))
		var buf bytes.Buffer
		if err := WriteFrame(&buf, kind, payload); err != nil {
			t.Fatalf("WriteFrame(%#x): %v", kind, err)
		}
		gotKind, gotPayload, err := ReadFrame(&buf)
		if err != nil {
			t.Fatalf("ReadFrame(%#x): %v", kind, err)
		}
		if gotKind != kind || !bytes.Equal(gotPayload, payload) {
			t.Fatalf("round-trip %#x: got (%#x,%q)", kind, gotKind, gotPayload)
		}
	}
}

// T4 (framing half): ReadFrame rejects an over-declared length BEFORE allocating — a
// worker cannot over-declare a giant length to OOM the key+DB supervisor. We feed ONLY
// the 4-byte prefix (no body); the ceiling must fire before any read of the body.
func TestReadFrameCeilingBeforeAlloc(t *testing.T) {
	var lenb [4]byte
	binary.BigEndian.PutUint32(lenb[:], 50<<20) // declare 50 MiB
	_, _, err := ReadFrame(bytes.NewReader(lenb[:]))
	if err == nil || !strings.Contains(err.Error(), "before alloc") {
		t.Fatalf("want ceiling-before-alloc error, got %v", err)
	}
	// A zero length is also rejected (need at least the kind byte).
	binary.BigEndian.PutUint32(lenb[:], 0)
	if _, _, err := ReadFrame(bytes.NewReader(lenb[:])); err == nil {
		t.Fatal("zero-length frame accepted")
	}
}

// K7 (proto v2, direction-split ceilings): WriteFrame is the sup→worker REQUEST writer, so it
// admits up to MaxRequestFrame (a multi-MB source image fits) but rejects beyond it. A former
// MaxFrame-sized payload is now ACCEPTED — the old single-ceiling assert broke on purpose (that
// break is the K7 red proof); this is the new contract.
func TestWriteFrameRequestCeiling(t *testing.T) {
	// A payload that fit the old MaxFrame ceiling is now well within the request ceiling: accepted.
	if err := WriteFrame(&bytes.Buffer{}, KindRenderRequest, make([]byte, MaxFrame)); err != nil {
		t.Fatalf("WriteFrame rejected a request within MaxRequestFrame: %v", err)
	}
	// One byte over MaxRequestFrame (total = payload+1) is rejected before allocation.
	if err := WriteFrame(&bytes.Buffer{}, KindRenderRequest, make([]byte, MaxRequestFrame)); err == nil {
		t.Fatal("WriteFrame accepted a payload over MaxRequestFrame")
	}
}

// K7 (T4 response wall preserved): the request ceiling must NOT leak into the response reader. A
// multi-MB length that sits BETWEEN MaxFrame and MaxRequestFrame — one a compromised worker could
// declare — is still rejected by ReadFrame before any allocation, so the split did not weaken the
// OOM isolation to the untrusted worker.
func TestReadFrameResponseStaysTight(t *testing.T) {
	if MaxRequestFrame <= MaxFrame {
		t.Fatalf("test premise: MaxRequestFrame (%d) must exceed MaxFrame (%d)", MaxRequestFrame, MaxFrame)
	}
	var lenb [4]byte
	binary.BigEndian.PutUint32(lenb[:], uint32(MaxFrame+1)) // multi-MB, over the tight response cap
	if _, _, err := ReadFrame(bytes.NewReader(lenb[:])); err == nil || !strings.Contains(err.Error(), "before alloc") {
		t.Fatalf("ReadFrame accepted a response over MaxFrame (T4 wall breached), got %v", err)
	}
}

func TestResponseFixedSize(t *testing.T) {
	raw := make([]byte, RawFrameSize)
	for i := range raw {
		raw[i] = byte(i)
	}
	// OK round-trip with the exact raw size.
	okMeta := ResponseMeta{V: 1, ID: "x", OK: true, W: 400, H: 300, Dither: "none"}
	enc, err := EncodeResponse(okMeta, raw)
	if err != nil {
		t.Fatalf("EncodeResponse ok: %v", err)
	}
	meta, gotRaw, err := DecodeResponse(enc)
	if err != nil || !meta.OK || !bytes.Equal(gotRaw, raw) {
		t.Fatalf("DecodeResponse ok: meta=%+v err=%v rawEq=%v", meta, err, bytes.Equal(gotRaw, raw))
	}

	// Encode rejects an OK response whose image != the fixed size (no compressed smuggle).
	if _, err := EncodeResponse(okMeta, raw[:RawFrameSize-1]); err == nil {
		t.Error("EncodeResponse accepted an undersized ok image")
	}
	// Error response must carry no image.
	errMeta := ResponseMeta{V: 1, ID: "x", OK: false, Err: &RenderErr{Kind: "throw", Msg: "boom"}}
	if _, err := EncodeResponse(errMeta, raw); err == nil {
		t.Error("EncodeResponse accepted an image on an error response")
	}
	encErr, err := EncodeResponse(errMeta, nil)
	if err != nil {
		t.Fatalf("EncodeResponse err-frame: %v", err)
	}
	if meta, gotRaw, err := DecodeResponse(encErr); err != nil || meta.Err == nil || len(gotRaw) != 0 {
		t.Fatalf("DecodeResponse err-frame: meta=%+v raw=%d err=%v", meta, len(gotRaw), err)
	}

	// Decode rejects a hand-forged OK payload whose trailing image is the wrong size.
	forged := make([]byte, 4+2+10) // metaLen=2, meta="{}", 10 image bytes (!= 360000)
	binary.BigEndian.PutUint32(forged[:4], 2)
	copy(forged[4:], "{}")
	if _, _, err := DecodeResponse(forged); err == nil {
		t.Error("DecodeResponse accepted an ok payload with a short image")
	}
}

// T2: the M4 request shape carries resolved secret VALUES only — never the secrets key,
// never a DB DSN. A structural scan of a marshalled request proves the leak surface is
// absent by construction.
func TestRequestValuesOnly(t *testing.T) {
	req := Request{
		V:  1,
		ID: "abc",
		Fn: RequestFn{ID: 42, Ver: 7, Source: "export default async()=>({})"},
		Ctx: RequestCtx{
			Serial: "S1", Channel: "living-room",
			Trigger: Trigger{Type: "render"}, Now: "2026-07-01T00:00:00Z",
		},
		Secrets:     map[string]string{"ha_token": "SECRET-VALUE-XYZ"},
		EgressAllow: []string{"ha.local:8123"},
		EgressCred:  "per-call-cred",
		Limits:      Limits{TimeoutMs: 8000, MemMB: 256},
	}
	blob, err := json.Marshal(req)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	js := string(blob)
	if !strings.Contains(js, "SECRET-VALUE-XYZ") {
		t.Error("resolved secret value missing from request")
	}
	for _, banned := range []string{"SECRETS_KEY", "SECRETS_KEY_PREV", "POSTGRES_", "dsn", "DATABASE_URL", "postgres://"} {
		if strings.Contains(js, banned) {
			t.Errorf("request leaked %q: %s", banned, js)
		}
	}
}
